mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp

```diff
@@ -0,0 +1,233 @@
+  constexpr static spirv::BuiltIn value{
+      spirv::BuiltIn::SubgroupLocalInvocationId};
+};
+
+template <typename OpTy>
+inline constexpr spirv::BuiltIn spirv_counterpart_builtin_v =
+    spirv_counterpart_builtin<OpTy>::value;
+
+/// Returns the result of creating an operation to get a reference to an element
+/// of a sycl::id or sycl::range.
+Value createGetOp(OpBuilder &builder, Location loc, Type dimMtTy, Value res,
+                  Value index, ArrayAttr argumentTypes,
+                  FlatSymbolRefAttr functionName) {
+  return TypeSwitch<Type, Value>(res.getType())
+      .Case<IDType, RangeType>([&](auto arg) {
+        // `this` type
+        using ArgTy = decltype(arg);
+        // Operation type depending on ArgTy
+        using OpTy = std::conditional_t<std::is_same_v<ArgTy, IDType>,
+                                        SYCLIDGetOp, SYCLRangeGetOp>;
+        return builder.create<OpTy>(
+            loc, dimMtTy, res, index, argumentTypes, functionName, functionName,
+            builder.getAttr<FlatSymbolRefAttr>(ArgTy::getMnemonic()));
+      });
+}
+
+/// Extracts component \p i from SPIRV composite \p values.
+Value getDimension(OpBuilder &builder, Location loc, Value values, int64_t i) {
+  return builder.create<spirv::CompositeExtractOp>(loc, values, i);
+}
+
+/// Converts n-dimensional operations of type \tparam OpTy to calls to \tparam
+/// builtin SPIR-V builtin.
+template <typename OpTy,
+          spirv::BuiltIn builtin = spirv_counterpart_builtin_v<OpTy>>
+class GridOpPattern : public OpConversionPattern<OpTy> {
+public:
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+
+protected:
+  constexpr static spirv::BuiltIn spirvBuiltin{builtin};
+};
+
+/// Replace \p op with a sequence of operations that:
+/// 1. Allocate a new array of the result type in the stack;
+/// 2. Initialize the dimensions of the array with the result of calling SPIR-V
+/// builtin \p builtin;
+/// 3. Load the value in position \p index (assumed to be inbounds).
+void rewriteNDIndex(Operation *op, spirv::BuiltIn builtin, Value index,
+                    const SPIRVTypeConverter &typeConverter,
+                    ConversionPatternRewriter &rewriter) {
+  // TODO: Get default dimensions from parent modules.
+  constexpr int64_t dimensions{3};
+  const auto values = spirv::getBuiltinVariableValue(
+      op, builtin, typeConverter.getIndexType(), rewriter);
+  const auto loc = op->getLoc();
+  const auto indexTy = rewriter.getIndexType();
+  const auto alloca = static_cast<Value>(rewriter.create<memref::AllocaOp>(
+      loc, MemRefType::get({dimensions}, indexTy)));
+  for (int64_t i = 0; i < dimensions; ++i) {
+    const auto c =
+        static_cast<Value>(rewriter.create<arith::ConstantIndexOp>(loc, i));
+    const auto val = static_cast<Value>(rewriter.create<arith::IndexCastOp>(
+        loc, indexTy, getDimension(rewriter, loc, values, i)));
+    rewriter.create<AffineStoreOp>(loc, val, alloca, c);
```

Comment:
```suggestion
  const Value alloca = rewriter.create<memref::AllocaOp>(
      loc, MemRefType::get({dimensions}, indexTy));
  for (int64_t i = 0; i < dimensions; ++i) {
    const Value c = rewriter.create<arith::ConstantIndexOp>(loc, i);
    const Value val = rewriter.create<arith::IndexCastOp>(
        loc, indexTy, getDimension(rewriter, loc, values, i));
    rewriter.create<AffineStoreOp>(loc, val, alloca, c);
```
I think that the static_cast<Value> can all be dropped if `alloca`, `c` and `val` are declared as Value (rather than auto).

Reply:
Semantics are equivalent, but I guess it's a matter of taste. I prefer explicitly exposing casts rather than "hiding" them. I've changed all of the occurrences anyway.